Use connection pooling

sibayanjasper opened the pull request.

sibayanjasper committed:
Always use connection pool

sibayanjasper committed:
Update README.md

wildtron commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -11,7 +11,31 @@
+        for _ in range(0, self.max_pool_size):
+            self.pool.put_nowait(Connection(self.conf))
+
+    def connect(self):
```
Shouldn't this check if the pool is already present?

sibayanjasper replied:
Is this necessary? `_initialize_pool` will already be called in `__init__` so it's sure that there is the pool.

wildtron commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -11,7 +11,31 @@
+        return con
+
+    def close(self, con):
+        con.close()
```
fails when `con` is not a `Queue`

sibayanjasper replied:
`con` is a `Connection` instance

wildtron commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -11,7 +11,31 @@
+
+class Connection(object):
+
+    def __init__(self, conf):
```
declarations inside this block will fail if `conf` does not have the required properties. Set a fallback in case?

ninnzz replied:
@sibayanjasper I agree with sherwin, set a default conf or catch the error.

sibayanjasper replied:
I already updated. I put the check in `DB.__init__` since it is the one that will be used and not the `Connection` class.

pandafox commented on README.md:
````diff
@@ -20,26 +20,33 @@ Minimal and straightforward when doing queries
 ```
-- Initialization: Accepts two parameters, first being the config object and the second specifying if autoconnect to db is enabled. If set to false, call `con.connect()` 
+
+- Initialization: Accepts two parameters, first being the config object and the second specifying the maximum number of connections in the pool (default is 20)
````
Two parameters still?

Also default to 10 connections

sibayanjasper replied:
Will we not be allowing users to set their desired maximum number of connections?

pandafox replied:
We will but the example you gave takes 1 parameter, no? The max connection should be defined in the config object?

db = nmi_mysql.DB(conf)

pandafox commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -11,7 +11,31 @@
+
+    def close(self, con):
+        con.close()
+        self.pool.put_nowait(con)
```
Shouldn't we check if the connection is managed by our pool first before closing it?

Additionally, shouldn't we remove the closed connection from the queue instead?

sibayanjasper replied:
I think we should put the connection in the queue again to be reused by other calls of `connect`.

pandafox replied:
In most connection pool, when a connection it is closed, you won't be able to reuse it (a new one need to be created), now I'm not sure if it is the same for pymysql.

But I guess since you require user to call connect themselves every time (instead of just get a pooled connection), this is fine.

A general guide of a pool connection:
1. user creates a connection pool
2. user just keep getting connection from the pool
2-1. user would be blocked if connection pool has no more connections
2-2. connection is returned to the pool when a query is finished (automatically)
3. user calls pool.close() when they don't need the connection anymore.

With this workflow the user never need to close/re-open the connection, otherwise, it defeats the purpose of connection pooling because each connection is re-established.

sibayanjasper replied:
I updated the implementation.
- When instantiating `DB`, it will create a connection pool with all `Connection` instances not yet connected.
- When executing a query, it will get a `Connection` instance from the pool, connect to the database, execute the query, close the connection, and return the `Connection` instance to the pool.

pandafox replied:
I like where this is going, but the connection should not be closed, the point of pooling is that the connection is being re-used so there's no need to create a new connection. If by close, you meant by release back to the pool (so that there's no need to login/create the connection again), then that's fine.

sibayanjasper replied:
Do you mean when creating the connection pool, all connections should be already open?
Is it okay to leave some connections hanging or idle until someone uses it? What if no one used it? What will happen?

pandafox replied:
1. No, the connection should be open if it was not already open (so we login/connect once and never have to do that again)
2. Usually the pool implementation would have a idle timeout, which would close the connection if its not used.

I thought the library would already support this, but if we need to write our own, I'd suggest we just switch to one that is well tested. There's no need to re-invent the wheel as this is actually quite hard to get it right.

sibayanjasper committed:
Check for invalid config object

sibayanjasper committed:
Raise exception on invalid config object

sibayanjasper committed:
Set max pool size to 10

sibayanjasper committed:
Check if connection to be returned is a Connection instance

sibayanjasper committed:
Include max_pool_size in config, update README.md

pprmint commented:
@sibayanjasper can you include tests of the pool working?
Also please include and update the documentation on how to use the pool. Thanks

sibayanjasper commented:
@pprmint the code already uses pooling by default. Also, the updated usage is in the README.

sibayanjasper committed:
Update nmi_mysql

Add method query in DB
Method query will get connection from the pool, execute the SQL query, close the connection, and return the connection to the pool

pandafox commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -9,9 +9,42 @@
+
+    def query(self, query, params=None):
+        con = self.pool.get()
+        con.connect()
```
This should only be done when con is not yet connected (which by default should already be connected if you get it from their pool?

For example: http://stackoverflow.com/questions/22699807/python-mysql-using-pymysql-auto-reconnect

Another thought, Instead of rewriting this, why don't we just use a proven library that has pooling support?

ex: 
http://www.sqlalchemy.org/
https://launchpad.net/pysqlpool
http://stackoverflow.com/questions/98687/what-is-the-best-solution-for-database-connection-pooling-in-python

sibayanjasper replied:
I talked to ninz and the reason why we built this is because pymysql lacks a good documentation. Also, we wanted to pattern it to anytv-node-mysql to achieve a more elegant code that will be easily used by the developers.
Also, for the libraries you suggested,
- pysqlpool only uses python2
- sqlalchemy is not simple since it's a big library (includes ORM) and you need more lines of code which we're avoiding since we want to have a clean and elegant code.

I'm thinking of changing this project's base to sqlalchemy instead of pymysql but it needs more time to be rewritten rather than continuing our pool implementation with this one.

pandafox replied:
I think all your analysis is spot on, and I understand why we did it. I also agree with your direction of rewriting the project based on sqlalchemy as it is one of the most used library (which means less bug/more stable/faster bug fixes etc). As for ORM, I agree that we should steer away from it as there's too much unnecessary abstraction (with different db engines for example) and limitations when it comes to querying, let's stay with sqlalchemy Core.

sibayanjasper committed:
Use sqlalchemy as base mysql client

sibayanjasper committed:
Use sqlalchemy parameter system

ninnzz commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -1,194 +1,122 @@
+        self.charset = 'utf8mb4'
+        self.con = None
+
+        self.engine = create_engine(
```
Is it assumed that it's the user's responsibility to catch the errors for the initialization?

sibayanjasper replied:
Possible errors for initialization is an invalid connection string which is already caught in checking config.
We don't need to worry about unexpected errors since we do not want to retry initialization for them. They should be the one to handle that.

ninnzz commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -1,194 +1,122 @@
-        self.max_pool_size = max_pool_size
-        self.initialize_pool()
+    def connect(self):
+        self.con = self.engine.connect()
```
put max retry :P
Its very useful especially when you are doing transactions.
Set the default retry to 5, then throw a max retry exception.
Refresh the max retry every successful transaction

pandafox replied:
Supporting retry is a good feature. However, I would disagree on default retry, it masks too many necessary error during development.

Please make sure retry is zero (ie no retry) unless the user specifies it.

ninnzz commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -1,194 +1,122 @@
+
         for param in _params:
+            if isinstance(param, list):
+                params.extend(param)
```
Hallow :P
Kindly document what this block of code is doing.
Other developers might not know the meaning of `.extend()`
Also, the whole code block is just flattening out the `_params` variable. Please indicate this in the comment.

ninnzz commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -1,194 +1,122 @@
+            if isinstance(param, list):
+                params.extend(param)
+
+            elif isinstance(param, dict):
```
I might be a little sleepy but does this code handle this case?
`db.query('UPDATE users SET %s WHERE name = %s', [{'name': 'ninz'}, 'jasper'])`

Will it not just return `UPDATE users SET ninz WHERE name = jasper`?

sibayanjasper replied:
It will become `SET name = %s` in method `_to_string`.

ninnzz commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -1,194 +1,122 @@
-    def return_connection(self, db):
-        return self.pool.put_nowait(db)
+        try:
+            result = self.con.execute(query, *params)
```
Can you add,
`self._latest_query` and `self._latest_param` variable?
Just to help in debugging.

pandafox replied:
Good idea, but can we name it _last_xxx instead?

ninnzz replied:
Lol, now that I think about it, `_last` is better. haha

sibayanjasper replied:
updated

pprmint commented:
@jvmsangkal, @sibayanjasper please include tests as well.
Test the ff:
-[ ] see if the old codes are working, do atleast two scripts/code that uses the old version
-[ ] test pool, see if it actually uses the pool
-[ ] error handling, give examples on how to handle different errors

pandafox commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -1,194 +1,122 @@
+
+    def _sql_alchemy_format(self, conf):
+        return ''.join([
+            'mysql+pymysql://',
```
Still uses pymysql?

sibayanjasper replied:
For `sqlalchemy` to support MySQL database, we need to select a dialect/DBAPI. See http://docs.sqlalchemy.org/en/latest/dialects/mysql.html#dialect-mysql.
The default DBAPI is MySQLdb which only supports Python 2.
By that, we'll still be needing to install a driver, in this case, `pymysql`.
Should we try the other ones or should we just stick with `pymysql`?

pandafox replied:
What about default python connector from mysql? Or do you guys have any bad experience with it? The official version seems to be a good version to use?

http://dev.mysql.com/downloads/connector/python/

sibayanjasper replied:
I tried python connector and it's working fine. Although I like pymysql since it is supported by pip, and is actively maintained and popular. But we can switch to python connector since it's the official Oracle-supported driver for MySQL.

pandafox replied:
Apparently, the performance of default mysql connector is a worse than pymysql, (why am I not surprised?...) so I guess we should continue using PyMySQL if we're just using it as underlying connector.

https://gist.github.com/methane/90ec97dda7fa9c7c4ef1
http://charlesnagy.info/it/python/python-mysqldb-vs-mysql-connector-query-performance
https://wiki.openstack.org/wiki/PyMySQL_evaluation

sibayanjasper replied:
Okay. Got it. :+1:

pandafox commented on README.md:
```diff
@@ -20,62 +20,88 @@ Minimal and straightforward when doing queries
-
-connection.close()
+data1 = db.query('SELECT * FROM mytable WHERE name = %s', ['ninz'])
+data2 = db.query('SELECT * FROM mytable WHERE name IN (%s) AND age = %s', [['john', 'doe'], 10])
```
Do we support multiple statement in one query? If so, can we give example (also state if we don't support this.)

sibayanjasper replied:
SQLAlchemy supports this but I don't recommend doing it. What do you think?

pandafox replied:
What's the reason you don't recommend doing it? I guess since we are doing pooling and the query are synchronous by nature it is not that big of a deal. More of a concern if we are in the async realm...

sibayanjasper replied:
For me, it's not good to the eyes. :) Also, it can be difficult to be tracked/debugged since multiple statements in one query will only return the first query results I think.
If we will not be concerned with the user in using this feature, then we can show some examples just to inform them that it's possible.

pandafox replied:
Hi Jasper,

Usually, for muti-statement queries, the results would be returned for each statement in an array. So there's really no loss in functionality.

As for the format/display/readability issue, that is more for the users to properly format the query themselves. a single statement query would also look horrible if everything is put in one line.

You can easily add enough spaces/SQL comments in the query if you want to make something more readable.

sibayanjasper replied:
Got it. I already updated the code.

It takes me time to fix support for multi-statement queries since sqlalchemy does not directly support this but pymysql does. See http://docs.sqlalchemy.org/en/latest/core/connections.html#multiple-result-sets.

Also, I'm having a problem on when to return affected rows since we already support multi-statement queries, I don't have a way to check each statement if it's a select or not. For now, I just return affected rows if there are no rows returned.

Here's the commit 4de0f558b573cf6480d5203190e565cbe7d72157.

sibayanjasper committed:
Update

Add retry to connect
Add _last_query and _last_params
Add comments

pandafox commented on nmi_mysql/nmi_mysql.py:
```diff
@@ -11,6 +11,7 @@
 
 CONFIG_KEYS = ['host', 'user', 'password', 'db', 'port']
 MAX_POOL_SIZE = 10
+MAX_RETRIES = 5
```
Default to 0, user should configure this per connection or even per query.

sibayanjasper replied:
MAX_RETRIES is used to catch large retry values (e.g. retry connecting 100 times which just does not make sense) or shouldn't we care about it?
The default retry value is 0. See this [line](https://github.com/pprmint/nmi_mysql/blob/cd00b1c89c7416617760ae79aaaf64b75802a71d/nmi_mysql/nmi_mysql.py#L104).

pandafox replied:
Ah, I got confused because MAX_POOL_SIZE and MAX_RETRIES are actually used differently. 
Let's make them do similar things?

Also, we shouldn't care if the user wants to set to 100 retries unless it breaks our library.

pandafox commented on README.md:
```diff
@@ -9,13 +9,14 @@ A very simple and intuitive mysql client wrapper for sqlalchemy.
-
 - Run the command to install: `pip install nmi_mysql`
-- Make sure you install [`sqlalchemy`](http://www.sqlalchemy.org/)
+- Make sure you install [`sqlalchemy`](http://www.sqlalchemy.org/) and [`pymysql`](http://www.pymysql.org/)
```
Please indicate what we use pymysql for, and include the reference I found as reference of making the decision vs mysql-connector

sibayanjasper committed:
Remove MAX_RETRIES and just use retry value from the user

sibayanjasper committed:
Add tests.py

sibayanjasper commented:
@pprmint 
- I included `tests.py` for testing.
- For pooling, we did not implement our own pooling since we used `sqlalchemy` which already supports pooling.
- For error handling, all expected errors that might arise are being caught. User errors should be handled by the user itself.

sibayanjasper committed:
Add autoconnect option

pandafox commented:
:dizzy:

sibayanjasper committed:
Add support for multi-statement queries

sibayanjasper committed:
Add executemany parameter to query() for bulk insert

sibayanjasper committed:
Add multi_query() for multi-statement queries

sibayanjasper committed:
Add note for not supporting tuple-parameterized INSERTs in multi-query

jvmsangkal commented:
Will this throw the same format of exception? or could you give an example on how to catch them?

sibayanjasper committed:
Raise err.orig rather than err

Existing projects use err.args to get the error code
err does not have args but err.orig does

sibayanjasper commented:
@jvmsangkal I updated the code and it will now raise `err.orig` rather than `err` only. There will now be no problem accessing args of exception.

jvmsangkal commented:
:star:
